core/lb.go

```diff
@@ -0,0 +1,202 @@
+		transcoder: lb.newT(transcoder, lb.workDir),
+		key:        key,
+		nonce:      nonce,
+		sender:     make(chan *transcoderParams, 1),
+	}
+	lb.sessions[job] = session
+	lb.load[transcoder] += costEstimate
+	lb.idx = (lb.idx + 1) % len(lb.transcoders)
+
+	// Local cleanup function
+	cleanupSession := func() {
+		lb.mu.Lock()
+		defer lb.mu.Unlock()
+		sess, exists := lb.sessions[job]
+		if !exists {
+			return
+		}
+		if sess.nonce != session.nonce {
+			// Session cached at `job` has since been recreated
+			return
+		}
+		delete(lb.sessions, job)
+		lb.load[transcoder] -= costEstimate
+		glog.V(common.DEBUG).Info("LB: Deleted transcode session for ", session.key)
+	}
+
+	go func() {
+		session.loop()
+		cleanupSession()
+	}()
+
+	glog.V(common.DEBUG).Info("LB: Created transcode session for ", session.key)
+	return session, nil
+}
+
+// Find the lowest loaded transcoder.
+// Expects the mutex `lb.mu` to be locked by the caller.
+func (lb *LoadBalancingTranscoder) leastLoaded() string {
+	min, idx := math.MaxInt64, 0
+	for i := 0; i < len(lb.transcoders); i++ {
+		k := (i + lb.idx) % len(lb.transcoders)
+		if lb.load[lb.transcoders[k]] < min {
+			min = lb.load[lb.transcoders[k]]
+			idx = k
+		}
+	}
+	return lb.transcoders[idx]
+}
+
+type transcoderParams struct {
+	job      string
+	fname    string
+	profiles []ffmpeg.VideoProfile
+	res      chan struct {
+		*TranscodeData
+		error
+	}
+}
+
+type transcoderSession struct {
+	transcoder TranscoderSession
+	key        string
+	nonce      string
+
+	sender chan *transcoderParams
+}
+
+func (sess *transcoderSession) loop() {
+	defer func() {
+		sess.transcoder.Stop()
+		sess.sender = nil // unblocks any sender + select{ ... }
+	}()
+	// Run everything on a single loop to mitigate threading issues,
+	//   especially around transcoder cleanup
+	for {
+		ctx, cancel := transcodeLoopContext()
+		select {
+		case <-ctx.Done():
+			// Terminate the session after a period of inactivity
+			glog.V(common.DEBUG).Info("LB: Transcode loop timed out for ", sess.key)
+			return
+		case params := <-sess.sender:
+			cancel()
+			res, err :=
+				sess.transcoder.Transcode(params.job, params.fname, params.profiles)
+			params.res <- struct {
+				*TranscodeData
+				error
+			}{res, err}
+			if err != nil {
+				glog.V(common.DEBUG).Info("LB: Stopping transcoder due to error for ", sess.key)
+				return
+			}
+		}
+	}
+}
+
+func (sess *transcoderSession) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+	params := &transcoderParams{job: job, fname: fname, profiles: profiles,
+		res: make(chan struct {
+			*TranscodeData
+			error
+		})}
+	// XXX check transcoding on a looped out channel . May be threading issues...
+	select {
+	case sess.sender <- params:
+		glog.V(common.DEBUG).Info("LB: Transcode submitted for ", sess.key)
+	default:
+		glog.V(common.DEBUG).Info("LB: Transcoder was busy; exiting ", sess.key)
+		return nil, ErrTranscoderBusy
+	}
+	res := <-params.res
```

Comment:
There is memory leak here. `sess.sender` is buffered (of size 1), so two different goroutines can make it to this line simultaneously. And if transcode for the first one will return error, this will make loop to exit, and thus second goroutine will be waiting at this line for eternity.

Reply:
Took the brute force route here and made a session-level lock for the whole function in https://github.com/livepeer/go-livepeer/pull/1124/commits/b9e09ffeac9470c420a0c489cb4df12f4bb2df8d

Reply:
If I understand it right, this still doesn't fix problem:
1. First `Transcode` invocation takes lock
2. Second invocation at the same time stop at lock and waits
3. loop receives error, sends it into `params.res` channel and starts exiting. At exit it stops and lock
4. Now we have two threads waiting at lock. First `Transcode` invocation releases lock, and if the second `Transcode` call takes it, then we have problem again.

Also there is problem in `case <-ctx.Done()` branch - it can lead to memory leak in the same way.
I think proper solution could be like this:
```golang
func (sess *transcoderSession) loop() {
	defer func() {
		sess.transcoder.Stop()
		sess.sender = nil // unblocks any sender + select{ ... }
	}()
	// Run everything on a single loop to mitigate threading issues,
	//   especially around transcoder cleanup
	for {
		ctx, cancel := sess.makeContext()
		select {
		case <-ctx.Done():
			// Terminate the session after a period of inactivity
			glog.V(common.DEBUG).Info("LB: Transcode loop timed out for ", sess.key)
			return
		case params := <-sess.sender:
			cancel()
			res, err :=
				sess.transcoder.Transcode(params.job, params.fname, params.profiles)
			params.res <- struct {
				*TranscodeData
				error
			}{res, err}
			if err != nil {
				glog.V(common.DEBUG).Info("LB: Stopping transcoder due to error for ", sess.key)
				return
			}
		}
	}
}

func (sess *transcoderSession) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
	params := &transcoderParams{job: job, fname: fname, profiles: profiles,
		res: make(chan struct {
			*TranscodeData
			error
		})}
	sess.mu.Lock()
	defer sess.mu.Unlock()
	select {
	case sess.sender <- params:
		glog.V(common.DEBUG).Info("LB: Transcode submitted for ", sess.key)
	default:
		glog.V(common.DEBUG).Info("LB: Transcoder was busy; exiting ", sess.key)
		return nil, ErrTranscoderBusy
	}
	select {
	case res := <-params.res:
		return res.TranscodeData, res.error
	case <-time.After(8 * time.Second):
		glog.V(common.DEBUG).Info("LB: Transcoder timeout; exiting ", sess.key)
		return nil, ErrTranscoderBusy
	}
}
```

Reply:
> loop receives error, sends it into params.res channel and starts exiting. At exit it stops and lock
Now we have two threads waiting at lock. First Transcode invocation releases lock, and if the second Transcode call takes it, then we have problem again.

I'm a little hesitant to add yet another layer of timeouts here, just because of how brittle it is (and we already have a number of other request-level timeouts to contend with). What might help this issue is to drain the channel after assigning `sess.sender = nil` in the `loop` cleanup function. Something like:

```golang
sess.mu.Lock()
sender := sess.sender
sess.sender = nil
sess.mu.Unlock()
// Drain the buffered channel
for {
  select {
  params := <-sender:
    params.res <- struct { *TranscodeData, error } { nil, ErrTranscodeEnded }
  default:
    return
  }
}
```

Reply:
Good idea, this should solve the problem.

Reply:
Fixed in 80d64d5b129d7e7ea6e4828d388ab499cf971dea . Hard to unit-test for directly without making testsuite-specific additions to the code, so I'll just leave things running under a stress test. Tightening this up seems to have removed the need for locking around the stub transcoder, which is a good thing. Note that we run all the load balancing tests via `-race` now so hopefully that helps a bit in terms of future-proofing thread safety here.

Reply:
Added unit test here 6e4d6981f69f349336385a8ba3f0e633379d9951